Implement mysql_clear_password

ldanilek opened the pull request:
Fixes https://github.com/launchbadge/sqlx/issues/2443 by implementing `AuthPlugin::MySqlClearPassword` as described in [mysql docs](https://dev.mysql.com/doc/dev/mysql-server/latest/page_protocol_connection_phase_authentication_methods_clear_text_password.html) and sufficient to connect to a MySQL server hosted by AWS using [IAM authentication with ephemeral passwords](https://docs.aws.amazon.com/AmazonRDS/latest/UserGuide/UsingWithRDS.IAMDBAuth.Connecting.AWSCLI.html).

For consistency with other clients, like [mysql](https://docs.rs/mysql/latest/mysql/struct.OptsBuilder.html#method.enable_cleartext_plugin), and to avoid accidental security issues, we also add a flag to ConnectionOpts to enable this plugin, as described [here](https://dev.mysql.com/doc/refman/8.0/en/cleartext-pluggable-authentication.html).

nipunn1313 commented on sqlx-mysql/src/protocol/connect/auth_switch.rs:
```diff
@@ -27,14 +27,23 @@ impl Decode<'_> for AuthSwitchRequest {
+            if matches!(plugin, AuthPlugin::MySqlClearPassword) {
+                // Contrary to the MySQL protocol, AWS Aurora with IAM sends
+                // no data. That is fine because the MySQL protocol says to
+                // ignore any data sent.
```
Can we link documentation that corroborates these comments? Like https://dev.mysql.com/doc/dev/mysql-server/latest/page_protocol_connection_phase_authentication_methods_clear_text_password.html
or
https://dev.mysql.com/doc/refman/8.0/en/cleartext-pluggable-authentication.html

If the `MySqlClearPassword` protocol wants you to ignore data - and we're making a specific exception for the aurora no-data scenario, perhaps we can make that more clear like this.

```
if matches!(plugin, AuthPlugin::MySqlClearPassword) && buf.is_empty() {
  // Contrary to the MySQL protocol, AWS Aurora with IAM sends no data.
  // In this scenario, it is ok to proceed with an empty data buffer.
  return Ok(Self { plugin, data: Bytes::new() })
}
```

And then have the rest of the code left the way it is.

nipunn1313 commented on sqlx-mysql/src/protocol/connect/auth_switch.rs:
```diff
@@ -27,14 +27,23 @@ impl Decode<'_> for AuthSwitchRequest {
+            let data = buf.get_bytes(20);
+            buf.advance(1); // NUL-terminator
+            data
+        };
```
IMO - this feels like a good candidate for a small unit test at the bottom of the file.

I see there's a similar example in sqlx-mysql/src/protocol/response/err.rs

nipunn1313 reviewed (changes requested):
nice!!

nipunn1313 commented:
looks good!

abonander commented:
@ldanilek `cargo fmt` needs to be run.

ldanilek commented:
@abonander i ran cargo fmt. can you take another look 🙏

abonander reviewed (changes requested):
I've been putting off merging this because honestly the idea is rather horrifying to me. 

With this plugin enabled, and without additional defensive configuration like `ssl-mode=VERIFY_IDENTITY` (which isn't even suggested in the AWS tutorial), it would be quite straightforward for an attacker, e.g. by a compromised router, to trick the application into divulging its credentials.

The documentation on the configuration option should at the very least recommend a stronger `.ssl_mode()` setting than the default, either `Required` or `VerifyCa` in the case of connecting to AWS where a root certificate is provided.

ldanilek commented:
> The documentation on the configuration option should at the very least recommend a stronger `.ssl_mode()` setting than the default, either `Required` or `VerifyCa` in the case of connecting to AWS where a root certificate is provided.

i could add an assertion on `ssl-mode` when `enable_cleartext_plugin` is true, if that would reduce the security risk. maybe in `DoHandshake::new` we could throw an error if `enable_cleartext_plugin && ssl_mode < REQUIRED`? does that sound like a good place to put such an assertion, and what kind of error would we throw?

abonander commented:
We shouldn't have a panic that you can't work around without enabling TLS as that could get really annoying.

I would accept maybe logging a warning, but making the implications clearer in the documentation is the main point.

ldanilek commented:
@abonander could you take another look please 🙏 . i've updated the comments to be scarier, and added a warning log when connecting in an insecure manner.
